include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,1221 @@
+GKO_REGISTER_MPI_TYPE(int, MPI_INT);
+GKO_REGISTER_MPI_TYPE(unsigned short, MPI_UNSIGNED_SHORT);
+GKO_REGISTER_MPI_TYPE(unsigned long, MPI_UNSIGNED_LONG);
+GKO_REGISTER_MPI_TYPE(long, MPI_LONG);
+GKO_REGISTER_MPI_TYPE(float, MPI_FLOAT);
+GKO_REGISTER_MPI_TYPE(double, MPI_DOUBLE);
+GKO_REGISTER_MPI_TYPE(long double, MPI_LONG_DOUBLE);
+GKO_REGISTER_MPI_TYPE(std::complex<float>, MPI_C_COMPLEX);
+GKO_REGISTER_MPI_TYPE(std::complex<double>, MPI_C_DOUBLE_COMPLEX);
+
+
+template <typename T>
+inline const T* in_place()
+{
+    return reinterpret_cast<const T*>(MPI_IN_PLACE);
+}
+
+
+/**
+ * This enum specifies the threading type to be used when creating an MPI
+ * environment.
+ */
+enum class thread_type {
+    serialized = MPI_THREAD_SERIALIZED,
+    funneled = MPI_THREAD_FUNNELED,
+    single = MPI_THREAD_SINGLE,
+    multiple = MPI_THREAD_MULTIPLE
+};
+
+
+/**
+ * Class that sets up and finalizes the MPI environment. This class is a simple
+ * RAII wrapper to MPI_Init and MPI_Finalize.
+ *
+ * MPI_Init must have been called before calling any MPI functions.
+ *
+ * @note If MPI_Init has already been called, then this class should not be
+ * used.
+ */
+class environment {
+public:
+    static bool is_finalized()
+    {
+        int flag = 0;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Finalized(&flag));
+        return flag;
+    }
+
+    static bool is_initialized()
+    {
+        int flag = 0;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Initialized(&flag));
+        return flag;
+    }
+
+    /**
+     * Return the provided thread support.
+     *
+     * @return the provided thread support
+     */
+    int get_provided_thread_support() { return provided_thread_support_; }
+
+    /**
+     * Call MPI_Init_thread and initialize the MPI environment
+     *
+     * @param argc  the number of arguments to the main function.
+     * @param argv  the arguments provided to the main function.
+     * @param thread_t  the type of threading for initialization. See
+     *                  @thread_type
+     */
+    environment(int& argc, char**& argv,
+                const thread_type thread_t = thread_type::serialized)
+    {
+        this->required_thread_support_ = static_cast<int>(thread_t);
+        GKO_ASSERT_NO_MPI_ERRORS(
+            MPI_Init_thread(&argc, &argv, this->required_thread_support_,
+                            &(this->provided_thread_support_)));
+    }
+
+    /**
+     * Call MPI_Finalize at the end of the scope of this class.
+     */
+    ~environment() { MPI_Finalize(); }
+
+    environment(const environment&) = delete;
+    environment(environment&&) = delete;
+    environment& operator=(const environment&) = delete;
+    environment& operator=(environment&&) = delete;
+
+private:
+    int required_thread_support_;
+    int provided_thread_support_;
+};
+
+
+/**
+ * Returns if GPU aware functionality has been enabled
+ */
+static bool is_gpu_aware()
+{
+#if GKO_HAVE_GPU_AWARE_MPI
+    return true;
+#else
+    return false;
+#endif
```

Comment:
I think this flag basically would control whether GPU Aware MPI is available and is taken directly from CMake and set by the user, so I think it should not differ between CUDA/HIP/DPC++

I removed the automatic detection, because that was not robust and making that consistent across different machine and backends is very difficult. So, the current behaviour is to just let the user specify if they have a GPU Aware MPI and compile/re-compile our MPI capabilities based on that.

Reply:
That might make it hard for people to run with MPI on heterogeneous systems with e.g. only CUDA-aware, but not ROCm-aware MPI? I would prefer removing the feature for now, until we see how to use GPU-aware MPI in practice? It seems pretty orthogonal to the rest